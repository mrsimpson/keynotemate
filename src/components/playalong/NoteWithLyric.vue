<template>
  <div
    class="note-with-lyric"
    :style="{ width: `${getDurationWidth(note.duration)}px` }"
  >
    <div class="note-container">
      <NoteBar :class="{ 'current-note': isCurrentNote }" :note="note" />
    </div>
    <div v-if="note.lyric" class="lyric">
      {{ note.lyric }}
    </div>
  </div>
</template>

<script setup lang="ts">
import { type ParsedNote } from '../../types/musicxml';
import NoteBar from './NoteBar.vue';
import { useNoteVisualizer } from '../../composables/useNoteVisualizer';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const props = defineProps<{
  note: ParsedNote;
  isCurrentNote?: boolean;
}>();

const { getDurationWidth } = useNoteVisualizer();
</script>

<style scoped>
.note-with-lyric {
  display: flex;
  flex-direction: column;
  gap: var(--spacing-sm);
  flex-shrink: 0;
}

.note-container {
  height: 60px; /* Fixed height for alignment */
  display: flex;
  align-items: flex-start; /* Align notes to top */
}

.lyric {
  font-size: 0.9rem;
  color: var(--text);
  text-align: center;
  max-width: 100%;
  overflow-wrap: break-word;
  min-height: 1.2em; /* Consistent height for lyrics */
}

:deep(.current-note) {
  transform: translateY(-10px);
  box-shadow: var(--shadow-lg);
  outline: 4px solid var(--primary);
}
</style>
