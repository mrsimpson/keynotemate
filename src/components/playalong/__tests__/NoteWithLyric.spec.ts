import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import type { ParsedNote } from '@/types/musicxml.ts'
import NoteWithLyric from '@/components/playalong/NoteWithLyric.vue'

describe('NoteWithLyric Component', () => {
  const mockNote: ParsedNote = {
    pitch: 'C',
    duration: {
      divisions: 4,
      type: 'quarter',
      relativeLength: 1,
    },
    octave: 4,
    lyric: 'Hello',
  }

  it('renders note with lyric snapshot', () => {
    const wrapper = mount(NoteWithLyric, {
      props: {
        note: mockNote,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders current note snapshot', () => {
    const wrapper = mount(NoteWithLyric, {
      props: {
        note: mockNote,
        isCurrentNote: true,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders note without lyric snapshot', () => {
    const noteWithoutLyric = { ...mockNote, lyric: undefined }
    const wrapper = mount(NoteWithLyric, {
      props: {
        note: noteWithoutLyric,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
