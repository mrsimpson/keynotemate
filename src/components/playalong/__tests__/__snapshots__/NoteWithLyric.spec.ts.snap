// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NoteWithLyric Component > renders current note snapshot 1`] = `
"<div data-v-abb92cc6="" class="note-with-lyric" style="width: 60px;">
  <div data-v-abb92cc6="" class="note-container">
    <div data-v-a4d447ea="" data-v-abb92cc6="" class="note-bar current-note" style="background-color: rgb(255, 0, 0); color: rgb(255, 255, 255); width: 60px; min-width: 40px;">
      <div data-v-a4d447ea="" class="note-content"><span data-v-a4d447ea="" class="note-label">C4</span></div>
    </div>
  </div>
  <div data-v-abb92cc6="" class="lyric">Hello</div>
</div>"
`;

exports[`NoteWithLyric Component > renders note with lyric snapshot 1`] = `
"<div data-v-abb92cc6="" class="note-with-lyric" style="width: 60px;">
  <div data-v-abb92cc6="" class="note-container">
    <div data-v-a4d447ea="" data-v-abb92cc6="" class="note-bar" style="background-color: rgb(255, 0, 0); color: rgb(255, 255, 255); width: 60px; min-width: 40px;">
      <div data-v-a4d447ea="" class="note-content"><span data-v-a4d447ea="" class="note-label">C4</span></div>
    </div>
  </div>
  <div data-v-abb92cc6="" class="lyric">Hello</div>
</div>"
`;

exports[`NoteWithLyric Component > renders note without lyric snapshot 1`] = `
"<div data-v-abb92cc6="" class="note-with-lyric" style="width: 60px;">
  <div data-v-abb92cc6="" class="note-container">
    <div data-v-a4d447ea="" data-v-abb92cc6="" class="note-bar" style="background-color: rgb(255, 0, 0); color: rgb(255, 255, 255); width: 60px; min-width: 40px;">
      <div data-v-a4d447ea="" class="note-content"><span data-v-a4d447ea="" class="note-label">C4</span></div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;
